inginious/frontend/environment_types/__init__.py

```diff
@@ -0,0 +1,31 @@
+import abc
+
+from inginious.common.hook_manager import HookManager
+from inginious.frontend.environment_types.docker import DockerEnvType
+from inginious.frontend.environment_types.mcq import MCQEnvType
+
+__env_types = {}
+
+
+def get_env_type(idx):
+    """ Return the FrontendEnvType with id idx. If this object does not exists, returns None."""
+    return __env_types.get(idx, None)
+
+
+def get_all_env_types():
+    return __env_types
+
+
+def register_env_type(env_obj):
+    """ env_obj is an instance of FrontendEnvType """
+    __env_types[env_obj.id] = env_obj
+
+
+def register_base_env_types(hook_manager: HookManager):
+    hook_manager.add_hook("register_env_type", register_env_type, 0)
+
+    # register standard env types here
+    register_env_type(DockerEnvType())
+    register_env_type(MCQEnvType())
```

Comment:
Does this mean the environnement are loaded in the frontend even if the agents are not launched ?

Reply:
Yes they are. For now this seems like a good middleground. In the long term, we may want that agents send themselves what they want as arguments, however this is complex to handle. Particularly when the frontend is starting or when an agent is not available. In that case, how would we be able to edit tasks, if we cannot get the parameters needed?

This architecture solves this problem, but forces the agent and the plugin "describing" the agent to be at the same version at all times.